xarray/core/indexing.py

```diff
@@ -437,16 +452,35 @@ def __init__(self, key):
         super().__init__(new_key)
 
 
-class ExplicitlyIndexed:
+T_ExplicitlyIndexed = TypeVar("T_ExplicitlyIndexed", bound="ExplicitlyIndexed")
+
+
+class ExplicitlyIndexed(metaclass=ABCMeta):
     """Mixin to mark support for Indexer subclasses in indexing."""
 
     __slots__ = ()
 
+    @abstractmethod
+    def __array__(self, dtype: DTypeLike = None) -> np.ndarray:
+        ...
+
+    @abstractmethod
+    def __getitem__(
+        self: T_ExplicitlyIndexed, key: Any
+    ) -> T_ExplicitlyIndexed | NumpyIndexingAdapter | np.ndarray | np.datetime64 | np.timedelta64:
+        ...
+        # TODO: fix the return type? Something more generic?
+
+    @property
+    @abstractmethod
+    def shape(self) -> tuple[int, ...]:
+        ...
 
-class ExplicitlyIndexedNDArrayMixin(NDArrayMixin, ExplicitlyIndexed):
+
+class ExplicitlyIndexedNDArrayMixin(NDArrayMixin, ExplicitlyIndexed, metaclass=ABCMeta):
```

Comment:
```suggestion
class ExplicitlyIndexedNDArrayMixin(NDArrayMixin, ExplicitlyIndexed, ABC):
```
last or first position?